README.md
# CSUwlan

CSUwlan是一个用于自动化登录数字中南电信网的Python程序，也附带有一些其他功能。

## 项目简介

作为CSU电信校园网用户，你是否已经厌倦了每次上网前要先打开浏览器并登录账号的一系列操作？CSUwlan即可代替你完成这些。它最初被设计于Windows环境中运行，现在已对Linux和Android（借助Termux）进行了适配，但受限于系统环境和作者的水平，此程序目前只能在Windows操作系统中**全自动的按需运行**。也就是说，在Windows中使用此程序，在连接到数字中南电信网无线网络后，你**无需任何操作**即可开始上网，而在其他操作系统中，你需要手动运行此程序以完成登录认证。

注：考虑到使用者的计算机可能并未安装Python解释器，故该项目中内置了可再发行的python_embed，且以Windows Shell Script编写了安装/卸载程序。另，该项目的编写者是由一位从未系统学习过编程的医学生，故代码质量可能不高。

## 运行依赖

+ Python依赖
    + Python解释器3.x 【已内置用于64位Windows系统的python_embed】
    + rsa模块（仅在配置登录密码，即直接运行config.py时依赖）【已内置于python_embed】
+ 运行于Windows时的依赖
    + Windows任务计划程序（用于在连接无线网络后自动启动CSUwlan程序）【Windows系统组件】
    + Windows Console Based Script Host（用于安装时解释vbs脚本以创建快捷方式）【Windows系统组件】
    + SCHTASKS、NETSH命令行工具（用于添加/删除任务计划，读取/控制WLAN连接）【Windows系统组件】
+ 运行于Linux时的依赖
    + ifconfig命令行工具（用于读取WLAN连接信息）<sup>[ 1](#脚注1)</sup>
+ 运行于Android时依赖
    + Termux应用
    + Termux-Api应用<sup>[ 1](#脚注1)</sup>
    + Termux中的python、termux-api<sup>[ 1 ](#脚注1)</sup>软件包
    
<a name="脚注1">1</a>: 可选，仅运行核心功能时无需

## 功能计划

+ [x] 登录数字中南电信网
+ [x] 登录信息历史记录
+ [x] 登出数字中南电信网<sup>[ 2](#脚注2)</sup>
+ [x] 回显WLAN接口信息并可按指定间隔刷新
+ [x] 通过Android Toast显示信息<sup>[ 3](#脚注3)</sup>
+ [x] 利用认证服务器的秘钥对密码进行RSA加密
+ [x] 独立的配置文件
+ [x] 包含安装程序
+ [x] 手动指定登录IP（可用于为其他设备登录）
+ [ ] 完善的安装/卸载程序

<a name="脚注2">2</a>: 仅在历史记录功能已启用（默认启用）时可用    
<a name="脚注3">3</a>: 依赖Termux-Api应用及软件包，且需同时传入`/print-to-android-toast`和`/nopause`两个命令行参数

## 安装及配置

### Windows

从[这个页面](https://github.com/zcx980605/CSUwlan/releases)下载带有Win64后缀的最新Release，解压到你希望的路径（不建议放置于“Program Files”目录<sup>[ 4](#脚注4)</sup>）后双击运行`.\CSUwlan\安装.bat`并按照提示完成安装。在安装过程中，你需要输入Windows当前用户的密码（如果没有则直接留空并按回车）用于添加任务计划，在最后一步，你需要输入数字中南电信网的账户名和密码（也就是之前从网页登录时要填的信息）。

如果你希望更改之前输入的数字中南电信网账户名或密码，请再次运行安装程序或直接运行`config.py`（推荐，如果可能），原有配置将备份为`config.py_bak.txt`。

### Linux

直接将此项目clone到本地，或从[这个页面](https://github.com/zcx980605/CSUwlan/releases)下载最新Release中的源代码并解压，后进入`./CSUwlan/app`目录，先执行`python3 config.py`来配置账户名和密码，完成即可后执行主程序`python3 AutoLogin.py`。示例：

```shell
git clone https://github.com/zcx980605/CSUwlan.git
cd CSUwlan/app
python3 config.py
python3 AutoLogin.py
```

如果你希望更改之前输入的数字中南电信网账户名或密码，请再次执行`python3 config.py`，原有配置将备份为`config.py_bak.txt`。

### Android

需要在Termux App中完成，具体操作与Linux终端相同。

### 详细配置（高级）

直接编辑安装程序生成的`config.py`（和`launcher_config.py`）可配置的参数有……

<a name="脚注4">4</a>: 登录历史记录数据库的默认位置位于安装目录，由于Windows系统的安全限制，程序只有取得管理员权限后才可对“Program Files”目录及子目录写入，此程序运行时不会请求管理员权限因而无法正常工作，除非您手动禁用历史记录或将历史记录数据库移动到其他位置（详见[详细配置](#详细配置)）。

## 如何使用

### 基本用法

+ 用于登录
    + 在Windows下，程序将于新连接到无线网络后自动运行，要手动运行，请直接双击桌面“登录数字中南”快捷方式图标，或不带参数的运行AutoLogin.py。
    + 在Linux下，你只能手动运行主程序来登录，即执行`python3 AutoLogin.py`。
    + 在Android下，建议安装Termux-Widget应用，以便启动器中添加指向脚本的图标来方便运行。你也可以直接Termux App中执行`python AutoLogin.py`来登录，具体操作与Linux终端相同。
+ 用于登出
    + 在Windows下，请直接双击桌面“登出数字中南”快捷方式图标，手动运行需向程序传入`/logout`参数，即`python AutoLogin.py /logout`。
    + 在Linux和Android下，请执行`python3 AutoLogin.py /logout`。

### 命令行参数详解

+ `/logout`：登出数字中南（使用历史记录数据库中最近的一条登录记录）
+ `/ip ip_addr`：指定登录IP（可用于为其他设备登录）。*ip_addr*为必选参数，有效值为“10.96.\[0-255\].\[0-255\]”。
+ `/sysinfo`：动态显示无线接口信息（默认刷新5次，间隔1.0秒）
    + 子开关`/r [[loops] [[interval]]]`：持续刷新，直到用户退出程序。可选指定循环次数和间隔，可单独指定循环次数，不可单独指定间隔。
+ `/print-to-android-toast`：将所有输出（包括标准输出、标准错误）重定向到Android Toast（仅在Android Termux环境下可用，且依赖Termux-Api应用及软件包）
+ `/nopause`：在本次运行时禁用类似“请按任意键继续”的提示，并在运行完成后直接退出，无论是否发生异常。
+ 内部参数`/location`：在Windows下，由`Launcher.py`传入，为用于获取登录Cookie的网址
+ 内部参数`/return`：在Windows下，由`Launcher.py`传入，指定成功登录并退出时返回的exitcode

## 如何卸载

卸载程序（脚本）还没有写😂，在Windows下，你需要手动删除名为“1_登录数字中南”的任务计划、手动删除桌面图标及手动删除程序所在的文件夹。

对于删除任务计划，你可以使用Windows管理工具中的任务计划程序（`taskschd.msc`），在GUI下删除“任务计划程序库”根目录的“1_登录数字中南”任务。也可以以管理员身份直接执行命令`schtasks /DELETE /TN "1_登录数字中南" /F`。

## 致谢

本项目引用了GitHub用户[@founddev](https://github.com/founddev)编写的[myrsa.py](https://github.com/founddev/everything/tree/master/rsa)。本项目的myrsa.py文件中亦标明了引用。

## 许可证

还没想好。。。目前NO LISENCE。


changelog.md
# 更新日志

First Version 0.9.0 @ 2019/05/01
  + 完成了核心功能

Version 0.9.1 @ 2019/05/04
  + 输出、错误处理

Version 0.9.2 @ 2019/05/11
  + 历史记录，超时硬限制

Version 0.10.0 @ 2019/08/25
  + 无需操作时不弹出窗口(构建了启动器Launcher.pyw)

Version 0.11.0 @ 2019/08/26
  + step3超时重试，命令行参数，模块化重构
  + 启动器调用主程序时利用命令行传参

Version 0.11.1 @ 2019/08/27
  + 登出功能

Version 0.11.2 @ 2019/09/01
  + 显示系统无线接口信息，重构主函数
  + 启动器中可指定python解释器路径、检测无线网络连接状态

Version 0.11.3 @ 2019/09/02
  + “账号已在线”错误时尝试注销并重试，启动器更新

Version 0.12.0 @ 2019/09/04
  + 独立的配置信息文件

Version 0.12.1 @ 2019/09/14
  + 启动器更新：阻止系统自动从无网络访问的WLAN断开(issue of Win10 1903)

Version 0.12.2 @ 2019/09/20
  + 配置文件支持直接运行以修改账号密码

Version 0.12.3 @ 2019/09/22
  + 启动器更新：无网络连接时在限定时间内重试

Version 0.12.4 @ 2019/09/28
  + 修改exitcode，使之符合Linux规范
  + 启动器更新：修改exitcode，线程同步(issue 20190926)

Version 0.12.5 @ 2019/10/20
  + 启动器更新：转义处理网址中的特殊含义字符以安全地回显消息(issue 20191011)

Version 0.13.0 @ 2020/01/10
  + 可利用termux-api输出到安卓Toast
  + 启动器更新：开始登陆同时即监测WLAN并阻止系统自动切换到其他网络而致step3、4失败(issue 20191026)

Version - @ 2021/11/15
  + 构建了安装脚本setup.py

Version 0.14.0 @ 2021/11/04
  + 可手动指定登录IP（用于为其他设备登录）
  + 其他修正

Version 0.14.1 @ 2021/11/05
  + 修正一处错误：（0.14.0中）传入location参数时未分析ip

- TODO issue 20200101 登陆成功后服务器未返回剩余流量等信息导致程序于存储历史记录前终止


app/myrsa.py
# 源代码：https://github.com/founddev/everything/tree/master/rsa
# 作者：（辣丝丝小白菜）https://www.52pojie.cn/home.php?mod=space&uid=706580
# 原文：https://www.52pojie.cn/forum.php?mod=viewthread&tid=874374
# 引用：2019/09/03

import rsa


class Encrypt(object):
    def __init__(self, e, m):
        self.e = e
        self.m = m

    def encrypt(self, message):
        mm = int(self.m, 16)
        ee = int(self.e, 16)
        rsa_pubkey = rsa.PublicKey(mm, ee)
        crypto = self._encrypt(message.encode(), rsa_pubkey)
        return crypto.hex()

    def _pad_for_encryption(self, message, target_length):
        message = message[::-1]
        max_msglength = target_length - 11
        msglength = len(message)

        padding = b''
        padding_length = target_length - msglength - 3

        for i in range(padding_length):
            padding += b'\x00'

        return b''.join([b'\x00\x00', padding, b'\x00', message])

    def _encrypt(self, message, pub_key):
        keylength = rsa.common.byte_size(pub_key.n)
        padded = self._pad_for_encryption(message, keylength)

        payload = rsa.transform.bytes2int(padded)
        encrypted = rsa.core.encrypt_int(payload, pub_key.e, pub_key.n)
        block = rsa.transform.int2bytes(encrypted, keylength)

        return block


def encrypt(e, m, message):
    en = Encrypt(e, m)
    return en.encrypt(message)


app/config.py
# Configuration file for "AutoLogin.py"

################################################################################
#### Note: This is an executable script. Any illegal modification may cause ####
#### application crash or function unexpectedly. BE CAREFUL WHEN EDITING!!! ####
################################################################################

# Timeout setting for network connection test (Step 1)
con_timeout = 5.0
hard_timeout = 5.0

# URL settings for network connection test (Step 1)
test_url = 'http://www.baidu.com/'
test_succeed_redirection = 'https://www.baidu.com/'

# URL pattern for "CSU ChinaNet" Network (Step 2)
url_pattern = 'http://61.137.86.87:8080/portalNat444/AccessServices/bas.59df7586?wlanuserip='

# Maximum retry times setting for getting cookie (Step 3)
step3_max_retry = 3

# Setting for whether pausing when application terminate
# Note: Use '/nopause' command argument to temporary disable pause
enable_pause = True

# Settings for history recording
# Note: Logout will not work if history recording is disabled
save_history = True
save_to_install_dir = True
# Be sure to uncomment the following line and assign a
# valid path if 'save_to_install_dir' is setted to 'False'
#history_db_dir = r'E:\Database'

# Login and Logout URL for "CSU ChinaNet" Network
login_url = 'http://61.137.86.87:8080/portalNat444/AccessServices/login'
logout_url = 'http://61.137.86.87:8080/portalNat444/AccessServices/logout?'

# Account settings for "CSU ChinaNet" Network
account = 'sample'
password = 'rsa'

# Settings for request header
ua = 'Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.198 Safari/537.36'
base_post_header = {
    'Host': '61.137.86.87:8080',
    'User-Agent': ua,
    'Accept': 'application/json, text/javascript, */*; q=0.01',
    'Accept-Language': 'zh-CN',
    'Accept-Encoding': 'gzip, deflate',
    'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8',
    'X-Requested-With': 'XMLHttpRequest',
    'Origin': 'http://61.137.86.87:8080',
    'Referer': 'http://61.137.86.87:8080/portalNat444/index.jsp',
    'Connection': 'keep-alive'
}

if __name__ == '__main__':
    import os
    import sys
    os.chdir(os.path.dirname(os.path.abspath(__file__)))
    sys.path.append(os.path.abspath(os.curdir))
    import myrsa

    print('\nAttention: Running this script may only modify account and password.')
    print('If you want to change other settings, please edit this file directly.')
    print('-'*30)
    new_account = input('account > ')
    new_account = new_account.strip()+'@zndx.inter'
    new_password = input('password> ')
    new_password = myrsa.encrypt('10001', 'a8a02b821d52d3d0ca90620c78474b78435423be99da83cc190ab5cb5b9b922a4c8ba6b251e78429757cf11cde119e1eacff46fa3bf3b43ef68ceb29897b7aa6b5b1359fef6f35f32b748dc109fd3d09f3443a2cc3b73e99579f3d0fe6a96ccf6a48bc40056a6cac327d309b93b1d61d6f6e8f4a42fc9540f34f1c4a2e053445', new_password)

    data = ''
    cur_file = os.path.abspath(__file__)
    bak_file = cur_file+'_bak.txt'
    with open(cur_file, 'r') as old:
        for line in old.readlines():
            if line[0:7] == 'account':
                line = "account = '"+new_account+"'\n"
            elif line[0:8] == 'password':
                line = "password = '"+new_password+"'\n"
            data += line
    if os.path.exists(bak_file):
        os.remove(bak_file)
    os.rename(cur_file, bak_file)
    with open(cur_file, mode='w+', encoding='utf-8') as new:
        new.writelines(data)
    print('Account information updated successfully.')
    print("Old settings backed up in file '%s'" % bak_file)
    input('\nPress Enter to continue...')


app/launcher_config.py
# Configuration file for "Launcher.pyw"

################################################################################
#### Note: This is an executable script. Any illegal modification may cause ####
#### application crash or function unexpectedly. BE CAREFUL WHEN EDITING!!! ####
################################################################################

# Timeout setting for network connection test (Step 1)
con_timeout = 5.0
hard_timeout = 5.0

# URL settings for network connection test (Step 1)
test_url = 'http://www.baidu.com/'
test_succeed_redirection = 'https://www.baidu.com/'

# URL pattern for "CSU ChinaNet" Network (Step 2)
url_pattern = 'http://61.137.86.87:8080/portalNat444/AccessServices/bas.59df7586?wlanuserip='

# Settings for request header
ua = 'Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.198 Safari/537.36'

# Path of python interpreter
path_of_python = r'E:\LinuxShared\CSUwlan\app\lib_python\python.exe'

# Time of monitoring WLAN interface after a succeed login
## Note: Win10 Version 1903 will automatically disconnect form 
##   the wireless access point that can not reach internet if 
##   user does not interact with the system notification related  
##   to it. This program will call Windows Net Shell (NETSH) and
##   reconnect to "CSU ChinaNet" access point if system close 
##   the connection in the following setted time.
wait_for_wlan_con_timeout = 20.0
monitor_wlan_if_timeout = 15.0


app/Autologin.py
import json
import os
import re
import sqlite3
import subprocess
import sys
import threading
import time
from urllib import error, parse, request

sys.path.append(os.path.dirname(os.path.abspath(__file__)))
try:
    from config import *
except ImportError:
    print("\nMissing configuration file 'config.py'. Please reinstall to fix.")
    time.sleep(2)
    sys.exit(2)
except SyntaxError as err:
    print("\nBad configuration file detected. Please correct.")
    print('\t'+str(err.args))
    time.sleep(2)
    sys.exit(255)

cur_time = time.time()
IsWindows = sys.platform == 'win32'
if IsWindows:
    os.system('TITLE 登录数字中南')
    if save_to_install_dir:
        os.chdir(os.path.dirname(os.path.abspath(__file__)))
    else:
        os.chdir(history_db_dir)

    def pause():
        if enable_pause:
            print('\n\n')
            os.system("PAUSE")

    def show_wlan_info(repeat, interval):
        i = 0 if repeat == 0 else 1
        interval = 0.0 if repeat == 1 else interval
        while i <= repeat:
            sub_proc = subprocess.Popen(
                'NETSH wlan show interfaces', shell=True, stdout=subprocess.PIPE)
            sub_proc.wait()
            wlan_sys_info = sub_proc.stdout.read().decode('gbk')
            os.system('CLS')
            print(wlan_sys_info)

            time.sleep(interval)
            if not repeat == 0:
                i = i+1

else:
    if save_to_install_dir:
        os.chdir(os.path.dirname(os.path.abspath(__file__)))
    else:
        os.chdir(history_db_dir)

    def pause():
        if enable_pause:
            input('\nPress Enter to continue...')

    def show_wlan_info(repeat, interval):
        i = 0 if repeat == 0 else 1
        interval = 0.0 if repeat == 1 else interval
        while i <= repeat:
            colums = os.get_terminal_size()[0]
            sub_proc = subprocess.Popen(
                'ifconfig wlan0', shell=True, stdout=subprocess.PIPE)
            sub_proc.wait()
            wlan_sys_info = sub_proc.stdout.read().decode('gbk')
            wlan_sys_info = wlan_sys_info.split('\n')
            flush_str = ''
            new_line = 0
            for item in wlan_sys_info:
                if len(item) <= colums:
                    flush_str = flush_str+item+' '*(colums-len(item))
                else:
                    flush_str = flush_str+item+' '*(len(item)-colums)
                    new_line = new_line+1
            sys.stdout.write(
                '\r\b'*(len(wlan_sys_info)+new_line)+flush_str+'\r')
            sys.stdout.flush()

            time.sleep(interval)
            if not repeat == 0:
                i = i+1


def handle_args():
    global enable_pause, location, ip_specified, skip_main, logout_mode, ret_code, print

    i = 1
    skip_main = False
    ip_specified = False
    logout_mode = False
    ret_code = 0
    toast_required = False
    arg_count = len(sys.argv)-1
    while i <= arg_count:
        argument = sys.argv[i]
        if argument == '/location':
            if i+1 <= arg_count:
                i = i+1
                location = sys.argv[i]
        elif argument == '/ip':
            if i+1 <= arg_count:
                i = i+1
                ip_specified = True
                verify_ip(sys.argv[i])
        elif argument == '/logout':
            skip_main = True
            logout_mode = True
        elif argument == '/sysinfo':
            skip_main = True
            repeat = 1
            interval = 1.0
            if (i+1 <= arg_count) and (sys.argv[i+1] == '/r'):
                i = i+1
                repeat = 0
                if (i+1 <= arg_count) and (sys.argv[i+1][0] != '/'):
                    i = i+1
                    repeat = int(sys.argv[i])
                if (i+1 <= arg_count) and (sys.argv[i+1][0] != '/'):
                    i = i+1
                    interval = float(sys.argv[i])
            show_wlan_info(repeat, interval)
        elif argument == '/nopause':
            enable_pause = False
        elif argument == '/return':
            if (i+1 <= arg_count) and (sys.argv[i+1][0] != '/'):
                i = i+1
                ret_code = int(sys.argv[i])
        elif argument == '/print-to-android-toast':
            if os.getenv('PREFIX') == None:
                print(
                    "The argument '"+argument+"'" + " can only be used in Termux. Aborted.\n")
                print('\n')
            else:
                if 'com.termux' in os.getenv('PREFIX'):
                    toast_required = True
                else:
                    print(
                        "The argument '"+argument+"'" + " can only be used in Termux. Aborted.\n")

        else:
            apperror(0, "Invalid argument - '"+argument+"'")
            sys.exit(22)
        i = i+1

    if toast_required:
        if enable_pause:
            print(
                "The argument '/print-to-android-toast' is valid only when '/nopause' assigned. Aborted.\n")
        else:
            def print(msg, cmd=''):
                cmd = ['termux-toast', '-b',
                       'red'] if cmd != '' else ['termux-toast', '-s']
                toast = subprocess.Popen(
                    cmd, stdin=subprocess.PIPE, universal_newlines=True)
                toast.communicate(input=str(msg))


def verify_ip(ip_to_verify=''):
    global ip

    ip_pattern = re.compile(
        '^10\.96\.(1\d{2}|2[0-4]\d|25[0-5]|[1-9]\d|\d)\.(1\d{2}|2[0-4]\d|25[0-5]|[1-9]\d|\d)$')
    if ip_pattern.match(ip_to_verify):
        ip = ip_to_verify
    else:
        apperror(0, "Invalid IP specified - '%s'" % ip_to_verify)


def try_req(req_obj, con_timeout_local=0.0, returnerr=False):
    global con_timeout
    if con_timeout_local == 0.0:
        con_timeout_local = con_timeout

    try:
        resp = request.urlopen(req_obj, timeout=con_timeout_local)
    except Exception as err:
        print(err)
        if returnerr:
            return err
    else:
        return resp


def apperror(step, reason, needpause=True):
    msg_list = ['Exception occured in step '+str(step)+':']
    msg_list.append('\t'+reason)
    print('\n'.join(msg_list), '\t')
    if needpause:
        pause()
    else:
        print('Exit in 3s...')
        time.sleep(3)
    sys.exit()


def write_history(start_time, data_dict):
    global ip_specified, cookie

    con = sqlite3.connect('csuwlan.db')
    cursor = con.cursor()
    rst = cursor.execute(
        "SELECT COUNT(*) FROM sqlite_master WHERE type='table' AND name='History'")
    if not rst.fetchall()[0][0] == 1:
        cursor.execute(
            '''CREATE TABLE History(
            Time DATETIME PRIMARY KEY NOT NULL,
            User VARCHAR(50) NOT NULL,
            IP VARCHAR(15) NOT NULL,
            IP_specified BOOLEAN NOT NULL,
            Total_flow NUMERIC(7,2) NOT NULL,
            Used_flow NUMERIC(7,2) NOT NULL,
            Surplus_flow NUMERIC(7,2) NOT NULL,
            Balance NUMERIC(5,2) NOT NULL,
            Update_time DATETIME NOT NULL,
            Cookie VARCHAR(255) NOT NULL);'''
        )

    part1 = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(cur_time))
    part2 = (cur_time - int(cur_time)) * 1000
    C1 = '%s.%03d' % (part1, part2)
    C2 = data_dict['accountID']
    C3 = data_dict['userIntranetAddress']
    C4 = ip_specified
    C5 = data_dict['totalflow']
    C6 = data_dict['usedflow']
    C7 = data_dict['surplusflow']
    C8 = data_dict['surplusmoney']
    C9 = data_dict['lastupdate']
    C10 = cookie
    values = [C1, C2, C3, C4, C5, C6, C7, C8, C9, C10]
    values = str(values)[1:-1]
    values = values.replace('True', '1')
    values = values.replace('False', '0')
    cursor.execute(
        "INSERT INTO History(Time, User, IP, IP_specified, Total_flow, Used_flow, Surplus_flow, Balance, Update_time, Cookie) " +
        "VALUES("+values+")"
    )
    con.commit()
    con.close()


def logout(post_header):
    con = sqlite3.connect('csuwlan.db')
    cursor = con.cursor()
    rst = cursor.execute(
        "SELECT COUNT(*) FROM sqlite_master WHERE type='table' AND name='History'")
    if rst.fetchall()[0][0] == 1:
        rst = cursor.execute("SELECT COUNT(*) FROM History")
        if rst.fetchall()[0][0] >= 1:
            rst = cursor.execute(
                "SELECT Time, User, IP, Cookie FROM History ORDER BY Time DESC LIMIT 0,1")
            records = rst.fetchall()[0]
            last_login_time = records[0]
            last_account = records[1]
            last_ip = records[2]
            last_cookie = records[3]
        else:
            apperror(1, 'No history record found.')
    else:
        apperror(1, 'Missing history record table.')

    post_header['Referer'] = 'http://61.137.86.87:8080/portalNat444/main2.jsp'
    post_header['Cookie'] = last_cookie
    post_data = 'brasAddress=59df7586&userIntranetAddress='+last_ip
    post_data = post_data.encode('utf-8')

    req = request.Request(url=logout_url, headers=post_header, data=post_data)
    resp = try_req(req, 10)

    if resp == None:
        apperror(2, 'Logout request posted, but failed to get response.')

    data_resp = json.loads(resp.read().decode('utf-8'))
    if data_resp['resultCode'] == '0':
        msg_list = ['Successfully logged out.\n', 'Connection info:']
        try:
            msg_list.append('%-17s%s' % ('  Account:', last_account))
            msg_list.append('%-17s%s' %
                            ('  Local IP:', data_resp['userIntranetAddress']))
            msg_list.append('%-17s%s' % ('  IP:', data_resp['userAddress']))
            msg_list.append('%-17s%s' % ('  Login time:', last_login_time))
            msg_list.append('%-17s%s' % ('  Duration:', data_resp['time']))
            msg_list.append('-'*43)
        except:
            print('\n'.join(msg_list))
            apperror(3, 'Missing required field. Info:\n\n'+str(data_resp_3))
        else:
            print('\n'.join(msg_list))
    else:
        apperror(3, 'Unknown Error.\n'+str(data_resp))


def internet_test():
    global test_url, hard_timeout
    global location

    class MyRedirectHandler(request.HTTPRedirectHandler):
        def http_error_302(self, req, fp, code, msg, hdrs):
            """重写父类http_error_302方法，阻止其自动跟踪重定向"""
            return fp

    class MyThread(threading.Thread):
        def __init__(self, target, args, daemon):
            """初始化父类属性和新增的report属性"""
            super().__init__(target=target, args=args, daemon=daemon)
            self.report = None

        def run(self):
            """重写父类run方法，使之可更新report属性"""
            try:
                if self._target:
                    self.report = self._target(*self._args, **self._kwargs)
            finally:
                del self._target, self._args, self._kwargs

    myhandler = MyRedirectHandler()
    myopener = request.build_opener(myhandler)
    request.install_opener(myopener)

    req_1 = request.Request(test_url)
    req_1.add_header('User-Agent', ua)
    while True:
        sub_thread = MyThread(try_req, (req_1,), True)
        sub_thread.start()
        sub_thread.join(hard_timeout)
        resp_1 = sub_thread.report
        if resp_1 == None:
            print(
                'Seems no connections established or network is unstable. System will retry in 3s.\n')
            time.sleep(3)
        else:
            break
    if resp_1.getcode() == 302:
        location = resp_1.getheader('location')
    else:
        apperror(1, 'Not the expected status code.')


def analyze_location():
    global url_pattern, test_succeed_redirection
    global location, ip

    if url_pattern in location:
        ip = location[location.find('wlanuserip=')+11:location.find('&')]
    elif location == test_succeed_redirection:
        apperror(1, 'Already have Internet access.', False)
    else:
        apperror(2, "Unsupported authentication page -'"+location+"'")


def get_cookie():
    global step3_max_retry, location
    global cookie

    req_2 = request.Request(location)
    req_2.add_header('User-Agent', ua)
    for ii in range(0, step3_max_retry):
        resp_2 = try_req(req_2)
        if resp_2 == None:
            if ii < step3_max_retry-1:
                print('Network is unstable. System will retry in 3s.\n')
                time.sleep(3)
            else:
                apperror(3, 'Reached maximum retry times. Exiting...')
        else:
            break
    if 'JSESSIONID=' in resp_2.getheader('Set-Cookie'):
        cookie = resp_2.getheader('Set-Cookie').split(';')[0]
    else:
        apperror(3, 'Failed to get cookie.')


def login(post_header, retry_when_unknown_err=False):
    global login_url, account, password, ip, cookie
    global data_resp_3
    global ret_code

    post_header['Cookie'] = cookie
    post_str = 'accountID='+account+'&password='+password + \
        '&brasAddress=59df7586&userIntranetAddress='+ip
    post_data = parse.quote(post_str, safe='=&').encode('utf-8')
    req_3 = request.Request(url=login_url, headers=post_header, data=post_data)

    resp_3 = try_req(req_3, 10, True)
    if isinstance(resp_3, BaseException):

        apperror(4, 'Login request posted, but failed to get response.')

    raw_resp_3 = resp_3.read().decode('utf-8')
    data_resp_3 = json.loads(raw_resp_3)
    if data_resp_3['resultCode'] == '0':
        msg_list = ['Successfully logged in.\n', 'Account info:']
        try:
            msg_list.append(
                '%-17s%s' % ('  Name:', data_resp_3['accountID']))
            msg_list.append(
                '%-17s%s' % ('  IP:', data_resp_3['userIntranetAddress']))
            msg_list.append(
                '%-17s%s' % ('  Total flow:', data_resp_3['totalflow']+'MB'))
            msg_list.append(
                '%-17s%s' % ('  Used flow:', data_resp_3['usedflow']+'MB'))
            msg_list.append(
                '%-17s%s' % ('  Surplus flow:', data_resp_3['surplusflow']+'MB'))
            msg_list.append(
                '%-17s%s' % ('  Balance:', '￥'+data_resp_3['surplusmoney']))
            msg_list.append('-'*43)
            msg_list.append(
                'Information updated at: ' + data_resp_3['lastupdate'])
        except:
            ret_code = 206
            print('\n'.join(msg_list))
            apperror(
                4, 'Missing required field. History writing Skipped. Info:\n\n'+str(data_resp_3))
        else:
            print('\n'.join(msg_list))
    elif data_resp_3['resultDescribe'] == '该账号已在线':
        msg_list = ['Exception occured in step 4:']
        msg_list.append(
            "\tServer returned error code ["+data_resp_3['resultCode'] + "]: '"+data_resp_3['resultDescribe']+"'\n")
        msg_list.append('Attempting to logout last recorded session....')
        print('\n'.join(msg_list))
        logout(base_post_header)
        print('Waiting for server refreshing account state.... 3s')
        time.sleep(3)
        print('Retrying login....')
        login(base_post_header, True)
    elif data_resp_3['resultDescribe'] == None and retry_when_unknown_err:
        print('Failed. Waiting for server refreshing account state.... 5s')
        time.sleep(5)
        print('\nRetrying login....')
        login(base_post_header, False)
    else:
        apperror(4, 'Unknown Error.\n\n'+str(data_resp_3))


def main():
    global ip, ip_specified, location, url_pattern
    handle_args()
    if skip_main:
        sub()
        return
    if ip_specified:
        location = url_pattern+ip
    if not 'location' in globals().keys():
        internet_test()
    if not ip_specified:
        analyze_location()
    get_cookie()
    login(base_post_header)
    if save_history:
        write_history(cur_time, data_resp_3)


def sub():
    if logout_mode:
        logout(base_post_header)


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        ret_code = 130
    except Exception as err:
        print(err, '\n')
        ret_code = 255
    pause()
    sys.exit(ret_code)
